### ISSUE TITLE
Welcome deverlopers on Windows

### ISSUE THREAD

**Issue by gagern** [CindyJS/CindyJS#107]
It would be nice if we could come up with a way how developers using Windows could build CindyJS from source. Since the code itself is JavaScript, being cross-platform seems natural, and it would be a shame if we lost that just because of a too *nix-oriented build system.

The main problem is the central Makefile and the utilities it relies on. We could ask users to install [Cygwin](http://cygwin.com/), but that's a pretty big install and not everyone might be comfortable with this. We could try [MSYS](http://mingw.org/wiki/MSYS) or [MSYS2](http://sourceforge.net/projects/msys2/), see whether that's good enough for our needs. Theoretically we could also try reproducing the functionality of GNU make using JavaScript. Or we could move away from Make ourselves, and use some JavaScript build tool (like [Gulp](https://www.npmjs.com/package/gulp) or [Grunt](https://www.npmjs.com/package/grunt) or something else) to control our builds.

The dependency on Java is probably something we won't get rid of easily. We need it to run the closure compiler and to build the GWT-based quickhull plugin. We can try to ensure that `make plain=1` (or its analogon in a new build system) works without Java. If we can get the tests to work as well, that would be great. And we could consider downloading Java automatically. I think it should be possible to set this up using environment variable tweaking, without permanent modifications to the system outside the CindyJS working directory.

**Comment by gagern** [comment]
Short status report: I've rewritten the build system using JavaScript. I wasn't happy with several of the concepts used by Gulp or Grunt. Grunt felt too inflexible for my taste, since we have some pretty exotic steps in there, and Gulp has the tendency to keep everything in memory, while the closure compiler reads its input from disk. Having the gulp task write them back to disk to a temporary dir doesn't make too much sense in my opinion. So I essentially duplicated the core logic of make: look at the inputs, look at the output(s), compare timestamps, and run tasks where the targets are outdated. Got some minor issue to work out yet, then a local test run on an emulated Windows, and then it should be ready for some public testing.

**Comment by kranich** [comment]
@gagern Can we close this?

**Comment by gagern** [comment]
I was waiting for feedback from @elkins0 on whether he can build on Windows. So far it's mostly theory, since the system on which I tested this has Cygwin installed, so it's not exactly your standard run of the line Windows with respect to what commands are available.

**Comment by elkins0** [comment]
Sorry I haven't looked at it yet. Has this changed since December 22, 2015? Do I need to be online to test this? Exactly what do you want tested? Do I need to be hooked up to github.com for this to work?

**Comment by gagern** [comment]
> Has this changed since December 22, 2015?

Yes, changes like those from #154 may affect you here. The #192 I just created might affect Windows builds as well, I think.

> Do I need to be online to test this?

Yes, that would be good since the build system will download stuff (like node modules and GWT) automatically. The GWT download in particular is over 100MB, so having a decent connection will help keep the waiting time low.

> Exactly what do you want tested?

Whether the build instructions in [the readme](https://github.com/CindyJS/CindyJS/blob/master/README.md#building) (a) work as advertised and (b) are clear enough to provide you with a `Cindy.js` built from source.

> Do I need to be hooked up to github.com for this to work?

I'm not exactly sure what you mean by “hooked up”. But I guess it depends. On the one hand, it would be interesting to have a build without even a git repository. To try that, just get the source code from a ZIP file. If you want to include #192 in the test even before it gets merged, get https://github.com/gagern/CindyJS/archive/nodevar.zip, otherwise https://github.com/CindyJS/CindyJS/archive/master.zip. To download that, you have to have access to the github web site, but you don't need to be logged in, and you definitely don't need any github-specific stuff installed or connected. A normal browser will do.

On the other hand, if you keep contributing the way you do, I imagine that having a local repository where you can commit to, and even commit changes to multiple files in a single commit, then having a git user interface of some kind installed would help, like e.g. [Github Desktop](https://help.github.com/desktop/guides/getting-started/), [Git for Windows](https://git-scm.com/download/win) or [SourceTree](https://www.sourcetreeapp.com/). We don't have CindyJS specific documentation on that, but there should be tutorials around, depending on the tool you choose.

I'd say we take steps one at a time: first check whether a build from ZIP works as intended, then see what links we might need to add to our documentation to help users set up Git as well.

**Comment by gagern** [comment]
>> Exactly what do you want tested?
>
> Whether the build instructions in [the readme](https://github.com/CindyJS/CindyJS/blob/master/README.md#building) (a) work as advertised and (b) are clear enough to provide you with a Cindy.js built from source.

If you continue contributing, then running the `alltests` build step would be useful as well. That's what Travis CI runs, so if you want to avoid complaints by Travis, you can run that locally and fix things before pushing the code to github. But one of the tests inside `alltests` is the `deploy` step, which in turn assumes that it's being executed from within a git repository since [it calls `git rev-parse HEAD`](https://github.com/CindyJS/CindyJS/blob/5be655ca6f42e9cfd45fb7873b542c2cba30bb9d/tools/prepare-deploy.js#L38) to determine the current commit id. So I'd say we test that when you have a working git setup.

**Comment by elkins0** [comment]
>>Exactly what do you want tested?

>Whether the build instructions in [the readme](https://github.com/CindyJS/CindyJS/blob/master/README.md#building) (a) work as advertised and (b) are clear enough to provide you with a `Cindy.js` built from source.

From the build instructions in [the readme](https://github.com/CindyJS/CindyJS/blob/master/README.md#building-on-windows):
>... So a standard release build would be node make build=release.
>
>Note that you should have the following software installed:
>
>    A recent Java Runtime Environment (JRE)
>    Node.js with the node command added to the PATH
>    Git for Windows with the git command usable from the Windows Command Prompt

Summary:
1. The versions of node and Git were not specified. The Makefile itself references node v0.12.6.
2. I used `git clone https://github.com/CindyJS/CindyJS`. [Not specified]
3. I chose to `CD /D C:\Temp\CindyJS` running node. [Not specified]
4. `node make build=release` worked as least as far as I could tell.

Details -- Here's what I did:

On a public server:

1. These files were downloaded:
```
01/27/2016  12:06 PM        29,629,384 PortableGit-2.7.0-64-bit.7z.exe
01/27/2016  12:17 PM         5,887,241 NodeJSPortable_5.3.0.paf.exe
01/27/2016  12:24 PM        30,702,384 Git-2.7.0-64-bit.exe
01/27/2016  12:53 PM        10,383,360 node-v5.5.0-x64.msi
```
2. Anything non-portable cannot installed on the public server due to security restrictions.
3. `PortableGit-2.7.0-64-bit.7z.exe` would not work on the public machine due to dependency on Microsoft.NET Framework 4.5 which could not installed.
4. No official portable version of is NodeJS available. So I used an un-official portable version from:
http://netcologne.dl.sourceforge.net/project/nodejsportable/v5.3.0/NodeJSPortable_5.3.0.paf.exe
5. Verified node from `NodeJSPortable_5.3.0.paf.exe` could be run.
6. Decided to try `Git-2.7.0-64-bit.exe` running from the home system to get the CindyJS repository before continuing.

Back on home system:

Installed `Git-2.7.0-64-bit.exe` and `node-v5.5.0-x64.msi`.

Ran `C:\Program Files\Git\git-bash.exe` and archived entire CindyJS directory created on 20160130.
```
$ git version
git version 2.7.0.windows.1
$ git clone https://github.com/CindyJS/CindyJS
Cloning into 'CindyJS'...
remote: Counting objects: 11571, done.
remote: Total 11571 (delta 0), reused 0 (delta 0), pack-reused 11571
Receiving objects: 100% (11571/11571), 11.33 MiB | 3.00 KiB/s, done.
Resolving deltas: 100% (7838/7838), done.
Checking connectivity... done.
Checking out files: 100% (737/737), done.
```
That took about just under an hour to complete.

Ran node three different times, with npm managing to crash the dial-up internet connection within 10 minutes each time.

Back on a public server:

1. Clicked on `NodeJSPortable_5.3.0.paf.exe` to create `C:\Temp\NodeJSPortable\*`.
2. Restored archive of CindyJS from `git clone https://github.com/CindyJS/CindyJS` into C:\Temp\CindyJS
3. Clicked on `NodeJSPortable\NodeJSPortable.exe` to bring up Command window
4. `CD /D C:\Temp\CindyJS`
5. `node make build=release Cindy.js`

Looks to me like `node make build=release Cindy.js` succeeded based on the contents of the build directory:
```
2016/02/03  10:40:43         0  CindyJS\build\node_modules.stamp
2016/02/03  10:40:47    637186  CindyJS\build\js\Cindy.plain.js
2016/02/03  10:40:47    860524  CindyJS\build\js\Cindy.plain.js.map
2016/02/03  10:40:55    313251  CindyJS\build\js\Cindy.closure.js
2016/02/03  10:40:55    820030  CindyJS\build\js\Cindy.closure.js.tmp.map
2016/02/03  10:40:56       138  CindyJS\build\prev-settings.json
2016/02/03  10:40:56    313243  CindyJS\build\js\Cindy.js
2016/02/03  10:40:56    820895  CindyJS\build\js\Cindy.js.map
2016/02/03  10:40:56    820903  CindyJS\build\js\Cindy.closure.js.map
```
Also created were:
```
2016/02/03  10:40:34     <DIR>  NodeJSPortable\Data\npm_cache\*; 519 Files; 682 Folders; 8.90 MB
2016/02/03  10:40:42     <DIR>  CindyJS\node_modules\*; 2645 Files; 478 Folders; 14.6 MB
2016/02/03  10:40:48   5654650  CindyJS\download\arch\compiler-20150126.zip
2016/02/03  10:40:48   6209423  CindyJS\download\closure-compiler\compiler-20150126.jar
```
See the [node-build-release-20160203.log](https://gist.github.com/elkins0/98d1a972cb8e0f35b955) for additional details.

**Comment by elkins0** [comment]
[Note: My home internet is now running at 12 Mbps]
@gagern : I'm stymied. I've tried this both from Git Shell and Git bash but when I try `node make build=release Cindy.js` the error is always the same--$NODE does not seem to be set as seen from npm:
```bash
Owner@Three MINGW64 ~/Documents/GitHub/gagern/CindyJS (conicDrawing2)
$ node make build=release Cindy.js
Running npm install to make sure we have all dependencies

> cindyjs@0.0.1 prepublish C:\Users\Owner\Documents\GitHub\gagern\CindyJS
> $NODE make/index.js call_npm=false parallel=true build=release all

'$NODE' is not recognized as an internal or external command,
operable program or batch file.

npm ERR! Windows_NT 6.1.7601
npm ERR! argv "C:\\Program Files\\nodejs\\node.exe" "C:\\Program Files\\nodejs\\node_modules\\npm\\bin\\npm-cli.js" "install"
npm ERR! node v5.5.0
npm ERR! npm  v3.3.12
npm ERR! code ELIFECYCLE
npm ERR! cindyjs@0.0.1 prepublish: `$NODE make/index.js call_npm=false parallel=true build=release all`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the cindyjs@0.0.1 prepublish script '$NODE make/index.js call_npm=false parallel=true build=release all'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the cindyjs package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     $NODE make/index.js call_npm=false parallel=true build=release all
npm ERR! You can get their info via:
npm ERR!     npm owner ls cindyjs
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     C:\Users\Owner\Documents\GitHub\gagern\CindyJS\npm-debug.log
C:\Users\Owner\Documents\GitHub\gagern\CindyJS\make\index.js:70
            throw Error("npm terminated with code " + code);
            ^

Error: npm terminated with code 1
    at Error (native)
    at ChildProcess.<anonymous> (C:\Users\Owner\Documents\GitHub\gagern\CindyJS\make\index.js:70:19)
    at emitTwo (events.js:100:13)
    at ChildProcess.emit (events.js:185:7)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
```
`npm-debug.log` shows:
```
68 verbose lifecycle cindyjs@0.0.1~prepublish: CWD: C:\Users\Owner\Documents\GitHub\gagern\CindyJS
69 silly lifecycle cindyjs@0.0.1~prepublish: Args: [ '/d /s /c',
69 silly lifecycle   '$NODE make/index.js call_npm=false parallel=true build=release all' ]
70 silly lifecycle cindyjs@0.0.1~prepublish: Returned: code: 1  signal: null
71 info lifecycle cindyjs@0.0.1~prepublish: Failed to exec prepublish script
72 verbose stack Error: cindyjs@0.0.1 prepublish: `$NODE make/index.js call_npm=false parallel=true build=release all`
72 verbose stack Exit status 1
72 verbose stack     at EventEmitter.<anonymous> (C:\Program Files\nodejs\node_modules\npm\lib\utils\lifecycle.js:232:16)
72 verbose stack     at emitTwo (events.js:100:13)
72 verbose stack     at EventEmitter.emit (events.js:185:7)
72 verbose stack     at ChildProcess.<anonymous> (C:\Program Files\nodejs\node_modules\npm\lib\utils\spawn.js:24:14)
72 verbose stack     at emitTwo (events.js:100:13)
72 verbose stack     at ChildProcess.emit (events.js:185:7)
72 verbose stack     at maybeClose (internal/child_process.js:821:16)
72 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:211:5)
73 verbose pkgid cindyjs@0.0.1
74 verbose cwd C:\Users\Owner\Documents\GitHub\gagern\CindyJS
```
My setup was:
```
md gagern
cd gagern
git clone https://github.com/gagern/CindyJS
cd  CindyJS
git checkout conicDrawing2
```
I've tried everything I know to put `NODE` into the environment, here in bash I used:
```
$ NODE=/c/Program\ Files/nodejs/node.exe
$ export NODE
$ echo $NODE
/c/Program Files/nodejs/node.exe
```
But no luck.  `node make build=release Cindy.js` seemed to work for me before...

**Comment by elkins0** [comment]
For now, looking at `package.json`:
```json
  "scripts": {
    "prepublish": "$NODE make/index.js call_npm=false parallel=true build=release all",
    "test": "$NODE make/index.js call_npm=false alltests"
  },
```
I have inferred that I can run these manually in `Git bash` presuming that the dependencies are satisfied:
```bash
node make/index.js call_npm=false parallel=true build=release all
node make/index.js call_npm=false alltests
```
The first tried to do something naughty (this requires super user access--why are we fiddling with java at all?) but appears to have succeeded otherwise:
```
[quickhull3d ] May 02, 2016 10:31:46 PM java.util.prefs.WindowsPreferences <init>
[quickhull3d ] WARNING: Could not open/create prefs root node Software\JavaSoft\Prefs at root 0x80000002. Windows RegCreateKeyEx(...) returned error code 5.
```
The tests passed. So I'll have a look at this tomorrow @gagern :
>I just had another look at this. I added the extra test cases from https://github.com/CindyJS/CindyJS/pull/280#issuecomment-209184409 but failed to reproduce the behavior demonstrated in the screen shot there. @elkins0, can you still demonstrate that behavior, if you enter the examples as you gave them into the file, and using the current CindyJS file from this branch? If so, what browser is that?

**Comment by gagern** [comment]
> I have inferred that I can run these manually in Git bash presuming that the dependencies are satisfied

Yes, well… the right way to ensure that dependencies are satisfied is by running `npm install`, which in turn will run the `prepublish` script. Getting the dependencies right without that step is tricky. One can use `CINDYJS_SKIP_PREPUBLISH` to avoid that step, i.e. run

    CINDYJS_SKIP_PREPUBLISH=true npm install

in Git bash or some other `sh`-like command interpreter, or probably something like

    SET CINDYJS_SKIP_PREPUBLISH=true
    npm install

in a pure `cmd.exe` Windows command line. That form would install the dependencies but stop after that.

> The first tried to do something naughty (this requires super user access

Whoa! That shouldn't be the case, in my opinion. The thing is that `node make/index.js all` does download things which it thinks are missing on the local machine. But it should download them into the `download` subdirectory of the working tree, and doing so should in my opinion not require super-user privileges. Can you provide more details, as to which process requires superuser privileges? What kind of message do you get if it doesn't have those privileges? Or was the message about `RegCreateKeyEx` the error you got when running this without extra privileges?

> why are we fiddling with java at all?

Because several build tools require Java. We use the Closure compiler to minify `Cindy.js`, and the advanced mode of that compiler for the Cindy3D and CindyGL plugins to provide some type checking and similar. We also use GWT to cross-compile the quickhull3d implementation from Java to JavaScript, although work is underway to replace that module with a pure Java version.

As to why Java is trying to create that registry key, [this Stack Overflow post](http://stackoverflow.com/q/16428098/1468366) provides some details, and [using a different preferences provider](http://www.davidc.net/programming/java/java-preferences-using-file-backing-store) might serve as a workaround. In any case, the warning should be harmless enough, so it's probably not worth the trouble.

> So I'll have a look at this tomorrow

Thanks, I appreciate that.

**Comment by elkins0** [comment]
@gagern wrote:
>>The first tried to do something naughty (this requires super user access

>Whoa! That shouldn't be the case, in my opinion. The thing is that  node make/index.js all  does download things which it thinks are missing on the local machine. But it should download them into the  download  subdirectory of the working tree, and doing so should in my opinion not require super-user privileges. Can you provide more details, as to which process requires superuser privileges? What kind of message do you get if it doesn't have those privileges? Or was the message about  RegCreateKeyEx  the error you got when running this without extra privileges?

I believe the popup message I saw requiring special access (it related to Java so I dismissed it) related to the `RegCreateKeyEx` change it tried to make during `node make/index.js call_npm=false parallel=true build=release all`:
```
[quickhull3d ] May 02, 2016 10:31:46 PM java.util.prefs.WindowsPreferences <init>
[quickhull3d ] WARNING: Could not open/create prefs root node Software\JavaSoft\Prefs at root 0x80000002. Windows RegCreateKeyEx(...) returned error code 5.
```
I'm not able to reproduce it now using this these commands in `Git bash`:
```bash
$ cd /c/Users/Owner/Documents/GitHub/gagern/CindyJS
$ npm install
> cindyjs@0.0.1 prepublish C:\Users\Owner\Documents\GitHub\gagern\CindyJS
> $NODE make/index.js call_npm=false parallel=true build=release all

'$NODE' is not recognized as an internal or external command,
operable program or batch file.
...
$ node make/index.js call_npm=false parallel=true build=release all
```

**Comment by kortenkamp** [comment]
As all current developers use Macs, I decided not to fix this (unless somebody complains, then I will reopen)
